pyccel/ast/cudatypes.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/devel/LICENSE for full license details.      #
+#------------------------------------------------------------------------------------------#
+""" Module containing types from the cuda module understood by pyccel
+"""
+from functools import lru_cache
+import numpy as np
+
+from pyccel.utilities.metaclasses import ArgumentSingleton
+
+from .datatypes     import FixedSizeNumericType, HomogeneousContainerType
+from .datatypes     import pyccel_type_to_original_type, original_type_to_pyccel_type
+
+from .numpytypes    import NumpyNDArrayType
+
+__all__ = ('CudaArrayType',)
+
+class CudaArrayType(HomogeneousContainerType, metaclass = ArgumentSingleton):
+    """
+    Class representing the Cuda array type.
+
+    Class representing the Cuda array type
+
+    Parameters
+    ----------
+    dtype : NumpyNumericType | PythonNativeBool | GenericType
+        The internal datatype of the object (GenericType is allowed for external
+        libraries, e.g. MPI).
+    rank : int
+        The rank of the new NumPy array.
+    order : str
+        The order of the memory layout for the new NumPy array.
+    memory_location : str
+        The memory location of the new cuda array ('host' or 'device').
+    """
+    __slots__ = ('_element_type', '_container_rank', '_order', '_memory_location')
+
+
+    def __init__(self, dtype, rank, order, memory_location):
+        assert isinstance(rank, int)
+        assert order in (None, 'C', 'F')
+        assert memory_location in ('host', 'device')
+
+        self._element_type = dtype
+        self._container_rank = rank
+        self._order = order
+        self._memory_location = memory_location
+        super().__init__()
+
+    @property
+    def memory_location(self):
+        """
+        The memory location of the new array ('host' or 'device').
+
+        The memory location of the new array ('host' or 'device').
+        """
+        return self._memory_location
+
+    @lru_cache
+    def __add__(self, other):
+        if(isinstance(other, CudaArrayType)):
+            assert self.memory_location == other.memory_location
```

Comment:
An assert is not the nicest way of testing this. Generally asserts are for Pyccel devs as they raise tracebacks. The `Errors` singleton could be used to raise a readable error to the user if they try to add 2 arrays on CPU and GPU

Reply:
A `NotImplementedError` is also caught and the messages are always displayed to users